### ISSUE TITLE
no SQ lines present in the header

### ISSUE THREAD

**Issue by zhangwenda0518** [marbl/Winnowmap#39]
To whom it may concern

In the process of using winnowmap, I encountered the following problems. Can you give me some help? thanks！

winnowmap -W repetitive_k9.txt -x map-pb -a -Y -L --eqx --MD --cs -t 60 ../hifi.fasta genome.fasta |samtools view -@ 20 -hb > read_alignment.bam

```
[M::mm_idx_gen::0.000*72.64] reading downweighted kmers
[M::mm_idx_gen::0.001*5.14] collected downweighted kmers, no. of kmers read=2270
[M::mm_idx_gen::0.001*5.12] saved the kmers in a bloom filter: hash functions=2 and size=32640
[M::mm_idx_gen::188.709*1.04] collected minimizers
[M::mm_idx_gen::202.242*1.07] sorted minimizers
**[WARNING] For a multi-part index, no @SQ lines will be outputted. Please use --split-prefix.**
[M::main::202.242*1.07] loaded/built the index for 234214 target sequence(s)
[M::main::202.242*1.07] running winnowmap in SV-aware mode
[M::main::202.242*1.07] stage1-specific parameters minP:1000, incP:1.99, maxP:8000, sample:1000, min-qlen:10000, min-qcov:0.5, min-mapq:5, mid-occ:5000
[M::main::202.242*1.07] stage2-specific parameters s2_bw:1000, s2_zdropinv:25
[M::mm_idx_stat] kmer size: 15; skip: 50; is_hpc: 0; #seq: 234214
[M::mm_idx_stat::202.337*1.07] distinct minimizers: 7146105 (31.16% are singletons); average occurrences: 22.090; average spacing: 25.344
[E::sam_parse1] **no SQ lines present in the header**
samtools view: error reading file "-"
[E::sam_parse1] no SQ lines present in the header
samtools view: error closing "-": -5
```
however，when I run with 
winnowmap -W repetitive_k9.txt -x map-pb **--split-prefix** -a -Y -L --eqx --MD --cs -t 60 ../hifi.fasta genome.fasta |samtools view -@ 20 -hb > read_alignment.bam
`[ERROR] --cs or --MD doesn't work with --split-prefix`
then I run whith out --cs or --MD
winnowmap -W repetitive_k9.txt -x map-pb --split-prefix -a -Y -L --eqx  -t 60 ../hifi.fasta genome.fasta |samtools view -@ 20 -hb > read_alignment.bam
`[main_samview] fail to read the header from "-".`

**Comment by zhangwenda0518** [comment]
yeah !
I run it success!

`winnowmap -k $kmer -W repetitive_k$kmer.txt -t 60 -ax map-pb --MD  genome.fasta ../hifi.fasta |samtools view -hb > alignment.bam`

**Comment by dazhuang420** [comment]
Hello,
I encountered the same issue you mentioned. The reason is that the SAM file from the winnowmap alignment is missing a header. I saw that you can convert the SAM file to a BAM file using samtools view -hb > alignment.bam. Can this BAM file be sorted afterward?
thanks

**Comment by RKeener** [comment]
Hello,
I have also encountered the same issue and found that if I specified a number of threads then it ran with no problem.
@dazhuang420 

**### TL;DR**
My original command that had problems: 
`winnowmap -W repetitive_k15.txt -ax map-ont hg002v1.0.1.arms.fasta 005.fq.gz  | samtools view -hb | samtools sort - > 005.bam
`
Command that ultimately worked: 
`winnowmap -W repetitive_k15.txt -ax map-ont hg002v1.0.1.arms.fasta 005.fq.gz -t 18 | samtools sort -o 005.bam  > 005.bam`

**### Full detail**
I subset my fastq files to increase processing speed and ran winnowmap over each of them. I only encountered this issue for 2 out of 35 files (file numbers 005 and 026) the others had no issue. 

For the sake of comparison I'll show analysis from a problematic file (005) and a file with no errors (001). 
Here are the summary stats on each input file: 
```
seqkit stats 005.fq.gz
file              format  type  num_seqs     sum_len  min_len  avg_len  max_len
005.fq.gz  FASTQ   DNA      9,128  88,994,896      501  9,749.7  327,785
seqkit stats 001.fq.gz
file              format  type  num_seqs     sum_len  min_len  avg_len  max_len
001.fq.gz  FASTQ   DNA      9,133  89,886,123      500  9,841.9   97,081

```
Here is my original winnowmap command: 
`winnowmap -W repetitive_k15.txt -ax map-ont hg002v1.0.1.fasta ${num}.fq.gz | samtools view -hb | samtools view - > ${num}.bam`

When I run this with file number 005 here is the output: 
```
[M::mm_idx_gen::0.055*0.06] reading downweighted kmers
[M::mm_idx_gen::0.078*0.06] collected downweighted kmers, no. of kmers read=2916
[M::mm_idx_gen::0.078*0.06] saved the kmers in a bloom filter: hash functions=2 and size=41928 
[M::mm_idx_gen::1.788*0.92] collected minimizers
[M::mm_idx_gen::1.913*0.96] sorted minimizers
[M::main::1.913*0.96] loaded/built the index for 96 target sequence(s)
[M::main::1.913*0.96] running winnowmap in SV-aware mode
[M::main::1.913*0.96] stage1-specific parameters minP:2000, incP:2.83, maxP:16000, sample:2000, min-qlen:10000, min-qcov:0.5, min-mapq:5, mid-occ:5000
[M::main::1.913*0.96] stage2-specific parameters s2_bw:2000, s2_zdropinv:25
[M::mm_idx_stat] kmer size: 15; skip: 50; is_hpc: 0; #seq: 96
[M::mm_idx_stat::1.923*0.96] distinct minimizers: 619431 (11.15% are singletons); average occurrences: 2.951; average spacing: 25.373
[main_samview] fail to read the header from "-".
samtools sort: failed to read header from "-"
```

Whereas when I run this with file number 001 (no issue) here is the output: 
```
[M::mm_idx_gen::0.001*3.53] reading downweighted kmers
[M::mm_idx_gen::0.002*2.11] collected downweighted kmers, no. of kmers read=2916
[M::mm_idx_gen::0.002*2.10] saved the kmers in a bloom filter: hash functions=2 and size=41928 
[M::mm_idx_gen::1.640*1.00] collected minimizers
[M::mm_idx_gen::1.762*1.04] sorted minimizers
[M::main::1.763*1.04] loaded/built the index for 96 target sequence(s)
[M::main::1.763*1.04] running winnowmap in SV-aware mode
[M::main::1.763*1.04] stage1-specific parameters minP:2000, incP:2.83, maxP:16000, sample:2000, min-qlen:10000, min-qcov:0.5, min-mapq:5, mid-occ:5000
[M::main::1.763*1.04] stage2-specific parameters s2_bw:2000, s2_zdropinv:25
[M::mm_idx_stat] kmer size: 15; skip: 50; is_hpc: 0; #seq: 96
[M::mm_idx_stat::1.772*1.04] distinct minimizers: 619431 (11.15% are singletons); average occurrences: 2.951; average spacing: 25.373
[M::worker_pipeline::132.270*2.70] mapped 9133 sequences
[M::main] Version: 2.03, pthreads=32, omp_threads=3
[M::main] CMD: winnowmap -W /data/abattle4/hji20/repetitive_k15.txt -ax map-ont /data/abattle4/hji20/hg002v1.0.1.arms.fasta ../mm2/001.fq.gz
[M::main] Real time: 132.288 sec; CPU: 357.741 sec; Peak RSS: 10.453 GB
```

Comparing the outputs, it looks like Winnowmap isn't completing on the problematic file but doesn't report why.

If I try to look at the output file I don't see anything for the 005.bam file: 
```
samtools view 005.bam | head
[main_samview] fail to read the header from "005.bam".
```

Whereas the same command with file 001.bam I see output in the expected format.

I asked a friend who also uses Winnowmap to try to process this file and they had no issue. We compared commands and the they specified the number of threads to use for winnowmap and a different samtools sort command.

When I updated my command to specify the number of threads: 
`winnowmap -W repetitive_k15.txt -ax map-ont hg002v1.0.1.arms.fasta 005.fq.gz -t 18 | samtools view -hb | samtools sort - > test.bam`

Winnowmap completed and output a .bam file.
The output was: 
```
[M::mm_idx_gen::0.013*0.25] reading downweighted kmers
[M::mm_idx_gen::0.014*0.32] collected downweighted kmers, no. of kmers read=2916
[M::mm_idx_gen::0.014*0.32] saved the kmers in a bloom filter: hash functions=2 and size=41928 
[M::mm_idx_gen::1.656*0.99] collected minimizers
[M::mm_idx_gen::1.789*1.02] sorted minimizers
[M::main::1.789*1.02] loaded/built the index for 96 target sequence(s)
[M::main::1.789*1.02] running winnowmap in SV-aware mode
[M::main::1.789*1.02] stage1-specific parameters minP:2000, incP:2.83, maxP:16000, sample:2000, min-qlen:10000, min-qcov:0.5, min-mapq:5, mid-occ:5000
[M::main::1.789*1.02] stage2-specific parameters s2_bw:2000, s2_zdropinv:25
[M::mm_idx_stat] kmer size: 15; skip: 50; is_hpc: 0; #seq: 96
[M::mm_idx_stat::1.799*1.02] distinct minimizers: 619431 (11.15% are singletons); average occurrences: 2.951; average spacing: 25.373
[M::worker_pipeline::122.043*2.72] mapped 9128 sequences
[M::main] Version: 2.03, pthreads=18, omp_threads=3
[M::main] CMD: winnowmap -W /data/abattle4/hji20/repetitive_k15.txt -ax map-ont -t 18 /data/abattle4/hji20/hg002v1.0.1.arms.fasta ../mm2/005.fq.gz
[M::main] Real time: 122.060 sec; CPU: 332.106 sec; Peak RSS: 7.919 GB
```
I tried the new command on my other problematic file and it also completed with no issue. 

However, when I used samtools to filter the files I got an error that they were truncated. 
`[main_samview] truncated file.`

I updated my command again to match my friend's command perfectly. My command was now: 
`winnowmap -W repetitive_k15.txt -ax map-ont hg002v1.0.1.arms.fasta 005.fq.gz -t 18 | samtools sort -o 005.bam  > 005.bam`
Now Winnowmap completes, a .bam file is output, and I can filter it using samtools with no errors. 
I will note that the reason my friend used the -o flag for samtools here is because they read (they thought maybe on one of Lior Pachter's twitter threads) that samtools is depreciating the -b command.

**Comment by RKeener** [comment]
I was able to solve my most recent comment. I updated the samtools command to also use threads but more importantly, it seems it was a memory issue. I previously discounted this because the code ran for me before on the same files, but our cluster was updated in between my analyses. 

Here was the final version of the command that worked for me: 
`winnowmap -W repetitive_k15.txt -ax map-ont ${reference} ${filename}.fq.gz -t 16 | samtools sort -o ${filename}.bam -@ 16 >> ${filename}.bam`

I looped this command over files that had around 3,000 reads each and reserved 1524 GB for my SLURM job.
